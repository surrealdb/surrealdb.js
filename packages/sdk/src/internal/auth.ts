import { SurrealError } from "../errors";
import type { AnyAuth } from "../types";

function convertString(
	auth: AnyAuth,
	result: Record<string, unknown>,
	from: string,
	to: string,
	optional?: boolean,
) {
	if (from in auth) {
		result[to] = `${auth[from as keyof AnyAuth]}`;
		delete result[from];
	} else if (optional !== true) {
		throw new SurrealError(
			`Key ${from} is missing from the authentication parameters`,
		);
	}
}

export function convertAuth(auth: AnyAuth): Record<string, unknown> {
	let result: Record<string, unknown> = {};

	if ("variables" in auth) {
		result = { ...auth.variables };
		convertString(auth, result, "access", "ac");
		convertString(auth, result, "namespace", "ns");
		convertString(auth, result, "database", "db");
	} else {
		convertString(auth, result, "access", "ac", true);
		convertString(auth, result, "database", "db", true);
		convertString(auth, result, "namespace", "ns", !("database" in auth));
		convertString(auth, result, "username", "user");
		convertString(auth, result, "password", "pass");
	}

	return result;
}
