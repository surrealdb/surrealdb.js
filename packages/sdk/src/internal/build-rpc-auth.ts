import { SurrealError } from "../errors";
import type { AnyAuth, ConnectionState } from "../types";

export function buildRpcAuth(state: ConnectionState, auth: AnyAuth): Record<string, unknown> {
    if ("key" in auth) {
        return {
            ns: auth.namespace,
            db: auth.database,
            ac: auth.access,
            key: auth.key,
        };
    }

    // Record user authentication
    if ("variables" in auth) {
        const namespace = auth.namespace ?? state.namespace;
        const database = auth.database ?? state.database;

        if (!database || !namespace) {
            throw new SurrealError(
                "Namespace and database must be provided or selected for record authentication",
            );
        }

        return {
            ...auth.variables,
            ac: auth.access,
            ns: namespace,
            db: database,
        };
    }

    // System authentication
    const access = "access" in auth ? auth.access : undefined;
    const namespace = "namespace" in auth ? auth.namespace : undefined;
    const database = "database" in auth ? auth.database : undefined;
    const result: Record<string, unknown> = {
        user: auth.username,
        pass: auth.password,
    };

    if (database && !namespace) {
        throw new SurrealError("Database authentication requires a namespace to be provided");
    }

    if (access) result.ac = access;
    if (namespace) result.ns = namespace;
    if (database) result.db = database;

    return result;
}
