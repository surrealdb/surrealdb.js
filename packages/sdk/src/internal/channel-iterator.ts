export class ChannelIterator<T> implements AsyncIterable<T>, AsyncIterator<T> {
    #cancelled = false;
    #queue: T[] = [];
    #waiter?: (r: IteratorResult<T>) => void;
    #cleanup?: () => void;

    constructor(cleanup?: () => void) {
        this.#cleanup = cleanup;
    }

    next(): Promise<IteratorResult<T>> {
        if (this.#cancelled) {
            return Promise.resolve({
                value: undefined,
                done: true,
            });
        }

        if (this.#queue.length > 0) {
            return Promise.resolve({
                value: this.#queue.shift() as T,
                done: false,
            });
        }

        return new Promise<IteratorResult<T>>((resolve) => {
            this.#waiter = resolve;
        });
    }

    return(): Promise<IteratorResult<T>> {
        this.#cancelled = true;
        this.#cleanup?.();
        this.#waiter?.({
            value: undefined,
            done: true,
        });

        return Promise.resolve({
            value: undefined,
            done: true,
        });
    }

    [Symbol.asyncIterator]() {
        return this;
    }

    submit(value: T): void {
        if (this.#cancelled) this.return;

        if (this.#waiter) {
            this.#waiter({ value, done: false });
            return;
        }

        this.#queue.push(value);
    }

    cancel(): void {
        this.#cancelled = true;
        this.#cleanup?.();
        this.#waiter?.({ value: undefined, done: true });
    }
}
