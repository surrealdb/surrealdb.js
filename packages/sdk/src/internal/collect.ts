import type { AnyRecordId } from "../types";
import { jsonify } from "../utils";
import { RecordId, StringRecordId } from "../value";
import { isAnyRecordId } from "./validation";

export type Output<T, S> = S extends AnyRecordId ? T : T[];

/** @deprecated */
export function output<T, S>(subject: S, input: T | T[]): Output<T, S> {
	if (subject instanceof RecordId || subject instanceof StringRecordId) {
		return (Array.isArray(input) ? input[0] : input) as Output<T, S>;
	}

	return (Array.isArray(input) ? input : [input]) as Output<T, S>;
}

export interface ProcessOptions {
	subject?: unknown;
	json?: boolean;
}

export function collect<R, I = unknown>(
	input: I | I[],
	options: ProcessOptions,
): R {
	let data: unknown = input;

	if (options.subject) {
		if (isAnyRecordId(options.subject)) {
			data = Array.isArray(input) ? input[0] : input;
		} else {
			data = Array.isArray(input) ? input : [input];
		}
	}

	if (options.json) {
		data = jsonify(data);
	}

	return data as R;
}
