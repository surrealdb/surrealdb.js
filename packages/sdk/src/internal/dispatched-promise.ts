type OnFulfilled<T, TResult> = ((value: T) => TResult | PromiseLike<TResult>) | null | undefined;
type OnRejected<TResult> = ((reason: unknown) => TResult | PromiseLike<TResult>) | null | undefined;

/**
 * A form of `Promise` that defers the computation of a Promise until it is
 * subscribed to. This allows for the configuration of the eventual `Promise`
 * before it is dispatched.
 */
export abstract class DispatchedPromise<T> extends Promise<T> {
	#resolve!: (value: T | PromiseLike<T>) => void;
	#reject!: (reason?: unknown) => void;
	#dispatched = false;

	protected abstract dispatch(): Promise<T>;

	constructor() {
		let _resolve: undefined | ((value: T | PromiseLike<T>) => void);
		let _reject: undefined | ((reason?: unknown) => void);

		super((resolve, reject) => {
			_resolve = resolve;
			_reject = reject;
		});

		if (!_resolve || !_reject) {
			throw new Error("resolve and reject required");
		}

		this.#resolve = _resolve;
		this.#reject = _reject;
	}

	#ensureDispatched(): Promise<T> {
		if (!this.#dispatched) {
			this.#dispatched = true;
			this.dispatch().then(this.#resolve, this.#reject);
		}

		return this;
	}

	override then<TResult1 = T, TResult2 = never>(
		onfulfilled?: OnFulfilled<T, TResult1>,
		onrejected?: OnRejected<TResult2>,
	): Promise<TResult1 | TResult2> {
		this.#ensureDispatched();
		return super.then(onfulfilled, onrejected);
	}

	override catch<TResult = never>(onrejected?: OnRejected<TResult>): Promise<T | TResult> {
		this.#ensureDispatched();
		return super.catch(onrejected);
	}

	override finally(onfinally?: (() => void) | undefined | null): Promise<T> {
		this.#ensureDispatched();
		return super.finally(onfinally);
	}

	static get [Symbol.species](): PromiseConstructor {
		return Promise;
	}

	[Symbol.toStringTag] = "DispatchedPromise";
}