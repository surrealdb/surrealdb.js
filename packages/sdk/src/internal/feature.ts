import {
    isVersionSupported,
    MAXIMUM_VERSION,
    MINIMUM_VERSION,
} from "../utils/is-version-supported";

export class Feature {
    #name: string;
    #since?: string;
    #until?: string;

    constructor(name: string, since?: string, until?: string) {
        this.#name = name;
        this.#since = since;
        this.#until = until;
    }

    get name(): string {
        return this.#name;
    }

    get sinceVersion(): string | undefined {
        return this.#since;
    }

    get untilVersion(): string | undefined {
        return this.#until;
    }

    supports(version: string): boolean {
        return isVersionSupported(
            version,
            this.#since ?? MINIMUM_VERSION,
            this.#until ?? MAXIMUM_VERSION,
        );
    }
}
