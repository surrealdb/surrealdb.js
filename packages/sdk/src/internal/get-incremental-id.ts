let id = 0;
export function getIncrementalID(): string {
	id = (id + 1) % Number.MAX_SAFE_INTEGER;
	return id.toString();
}
