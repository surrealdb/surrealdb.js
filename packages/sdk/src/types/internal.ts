import type { Jsonify } from "../utils";

export type MaybeJsonify<T, J extends boolean> = J extends true ? Jsonify<T> : T;
export type Prettify<T> = { [K in keyof T]: T[K] } & {};
