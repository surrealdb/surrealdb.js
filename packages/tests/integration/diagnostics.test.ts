import { beforeEach, describe, expect, test } from "bun:test";
import { applyDiagnostics, createRemoteEngines, type Diagnostic, RecordId } from "surrealdb";
import { resetIncrementalID } from "../../sdk/src/internal/get-incremental-id";
import { type Person, setupServer } from "./__helpers__";

const { createSurreal } = await setupServer();

beforeEach(() => {
    resetIncrementalID();
});

describe("diagnostics", async () => {
    test("diagnostic events", async () => {
        const events: Diagnostic[] = [];

        const surreal = await createSurreal({
            driverOptions: {
                engines: applyDiagnostics(createRemoteEngines(), (event) => {
                    // Skip the version event for test integrity
                    if (event.type === "version" && event.phase === "after" && event.success) {
                        return;
                    }

                    events.push(event);
                }),
            },
        });

        await surreal.version();

        resetIncrementalID();

        await surreal.create<Person>(new RecordId("person", 1)).content({
            firstname: "John",
            lastname: "Doe",
        });

        expect(events).toMatchSnapshot();
    });

    test("extract query", async () => {
        let query = "";
        let params = {};

        const surreal = await createSurreal({
            driverOptions: {
                engines: applyDiagnostics(createRemoteEngines(), (event) => {
                    if (event.type === "query" && event.phase === "after" && event.success) {
                        query = event.result.query;
                        params = event.result.params;
                    }
                }),
            },
        });

        await surreal.create<Person>(new RecordId("test", "test")).content({
            firstname: "John",
            lastname: "Doe",
        });

        expect(query).toMatchSnapshot();
        expect(params).toMatchSnapshot();
    });
});
