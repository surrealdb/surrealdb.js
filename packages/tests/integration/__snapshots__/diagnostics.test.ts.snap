// Bun Snapshot v1, https://bun.sh/docs/test/snapshots

exports[`diagnostics diagnostic events 1`] = `
[
  {
    "key": Uuid {},
    "phase": "before",
    "type": "open",
  },
  {
    "duration": Duration {},
    "key": Uuid {},
    "phase": "after",
    "result": undefined,
    "success": true,
    "type": "open",
  },
  {
    "key": Uuid {},
    "phase": "before",
    "type": "version",
  },
  {
    "key": Uuid {},
    "phase": "before",
    "type": "use",
  },
  {
    "duration": Duration {},
    "key": Uuid {},
    "phase": "after",
    "result": {
      "requested": {
        "database": "test",
        "namespace": "test",
      },
      "session": undefined,
    },
    "success": true,
    "type": "use",
  },
  {
    "key": Uuid {},
    "phase": "before",
    "type": "signin",
  },
  {
    "duration": Duration {},
    "key": Uuid {},
    "phase": "after",
    "result": {
      "session": undefined,
      "variant": "system_user",
    },
    "success": true,
    "type": "signin",
  },
  {
    "key": Uuid {},
    "phase": "before",
    "type": "version",
  },
  {
    "key": Uuid {},
    "phase": "before",
    "type": "query",
  },
  {
    "key": Uuid {},
    "phase": "progress",
    "result": {
      "chunk": {
        "batch": 0,
        "kind": "single",
        "query": 0,
        "result": [
          {
            "firstname": "John",
            "id": RecordId {},
            "lastname": "Doe",
          },
        ],
        "stats": {
          "bytesReceived": -1,
          "bytesScanned": -1,
          "duration": Duration {},
          "recordsReceived": -1,
          "recordsScanned": -1,
        },
        "type": undefined,
      },
      "params": {
        "bind__5": RecordId {},
        "bind__6": {
          "firstname": "John",
          "lastname": "Doe",
        },
      },
      "query": "CREATE ONLY $bind__5 CONTENT $bind__6",
      "session": undefined,
      "transaction": undefined,
    },
    "type": "query",
  },
  {
    "duration": Duration {},
    "key": Uuid {},
    "phase": "after",
    "result": {
      "params": {
        "bind__5": RecordId {},
        "bind__6": {
          "firstname": "John",
          "lastname": "Doe",
        },
      },
      "query": "CREATE ONLY $bind__5 CONTENT $bind__6",
      "session": undefined,
      "transaction": undefined,
    },
    "success": true,
    "type": "query",
  },
]
`;

exports[`diagnostics extract query 1`] = `"CREATE ONLY $bind__4 CONTENT $bind__5"`;

exports[`diagnostics extract query 2`] = `
{
  "bind__4": RecordId {},
  "bind__5": {
    "firstname": "John",
    "lastname": "Doe",
  },
}
`;
